MPS files when a daily optimization is asked

guilpier-code opened the pull request:
Before this fix, when user asked for a daily optimization, only 2 MPS files used to be created per week (**problem-\<year\>-\<week\>--optim-nb-\<opt-number\>.mps**) .
This is because MPS files generated by some daily optimization were overridden by files generated by the next daily optimization : indeed, they had the same name.
Now, MPS files wear the name : **problem-\<year\>-\<week\>-\<day\>--optim-nb-\<opt-number\>.mps** so they cannot be overridden.
Note that what we say here about MPS files is also true for criterion files (ex : **criterion-1-2-6--optim-nb-1.txt**)

Besides, some preliminary refactoring works were made to ease the current work. As usual, associated branches are sitting on top of each other : 
- #1030 
- #1032
- #1040

guilpier-code committed:
Moving current year and week from study to weekly optimization problem

Why ?
- Because weekly problem is more the location of current year and week (the problem knows what year and week it's solving)
- It makes the code cleaner (reduces dependency on the study)
- It makes easier some changes to come (export MPS files when optimization is daily)

Some collateral cleaning was made

guilpier-code committed:
move year and week from study to weekly opt problem : correction after sonar Q review

guilpier-code committed:
move year and week from study to weekly opt problem : correction after human review

guilpier-code committed:
Refactor : about optimization number

- Information of optimization number is spread at 2 locations : it can be found inside ProblemeHebdo and inside ProblemeHebdo->ProblemeAResoudre
- This piece of information does not need to be nested in these structures
So it is declared and defined once, and locally, and then passed to lower functions when needed.

guilpier-code committed:
Refactor : about optimization number : oops, correction of a unexpected remove

guilpier-code committed:
Refactor : about optimization number : removing unnecessary trailing affectation

guilpier-code committed:
MPS when daily optimization : code simplifications

- unnecessary argument in the mpsWriterFactory's constructor
- we don't need to test optNumber in function that build the MPS and criterion file names

guilpier-code committed:
MPS when daily optimization : code now produces 14 MPS files per week

guilpier-code committed:
MPS when daily optimization : adding "override" key words

guilpier-code committed:
MPS when daily optimization : in MPS files' name, change the order : day-week-year ==> year-day-week

guilpier-code committed:
MPS when daily optimization : corrections after sonar Q (and human) review

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,57 @@
+    unsigned int year_ = 0;
+};
+
+std::shared_ptr<optPeriod> createOptimizationPeriod(bool isOptimizationWeekly,
```
```suggestion
std::shared_ptr<optPeriod> createOptimizationPeriod(Antares::Data::SimplexOptimization optPeriod,
```

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,57 @@
+// ------------------------------------
+// Optimization period : base class
+// ------------------------------------
+class optPeriod
+{
```
// This class intends to provide an interface for the naming of optimization-related files (MPS files, etc.)

guilpier-code replied:
Done

flomnes commented on src/solver/utils/mps_utils.cpp:
```diff
@@ -324,58 +321,41 @@ void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(
 {}
 
-void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer)
+void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, std::string filename)
```
```suggestion
void fullMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, const std::string& filename)
```

guilpier-code replied:
Done

flomnes commented on src/solver/utils/mps_utils.cpp:
```diff
@@ -324,58 +321,41 @@ void OPT_EcrireJeuDeDonneesLineaireAuFormatMPS(
 {
 }
-void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer)
+void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, std::string filename)
```
```suggestion
void fullOrToolsMPSwriter::runIfNeeded(Solver::IResultWriter::Ptr writer, const std::string& filename)
```

guilpier-code replied:
Done

flomnes commented on src/solver/utils/mps_utils.h:
```diff
@@ -17,10 +17,8 @@ using namespace Antares::Optimization;
-                                               uint,
-                                               Solver::IResultWriter::Ptr writer);
+                                               Solver::IResultWriter::Ptr writer,
+                                               std::string filename);
```
```suggestion
                                               const std::string& filename);
```

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -243,11 +244,14 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo,
-
-    mpsWriterFactory mps_writer_factory(
-      ProblemeHebdo, NumIntervalle, numeroOptimisation, &Probleme, ortoolsUsed, solver);
+    std::string filename = createMPSfilename(opt_period, numeroOptimisation);
```
```suggestion
    const std::string filename = createMPSfilename(opt_period, numeroOptimisation);
```

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/opt_fonctions.h:
```diff
@@ -60,7 +61,7 @@ bool OPT_PilotageOptimisationQuadratique(PROBLEME_HEBDO*);
 ** \return True si l'operation s'est bien deroulee, false si le probleme n'a pas de solution
 */
-bool OPT_AppelDuSimplexe(PROBLEME_HEBDO*, int, const int);
+bool OPT_AppelDuSimplexe(PROBLEME_HEBDO*, int, const int, std::shared_ptr<optPeriod>);
```
Add explanations on this new argument (daily/weekly opt.)

guilpier-code replied:
Done

guilpier-code committed:
[skip ci] MPS when daily optimization : add comments on optimization period (as string)

guilpier-code committed:
MPS when daily optimization : logging a MPS / criterion filename only if we actually print the associated file

flomnes committed:
Merge remote-tracking branch 'github/fix/refactor-export-structure' into fix/mps-files-when-daily-opt

flomnes committed:
Fix & add tests related to MPS files (simplex-range = day, week)

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,74 @@
+                                                    unsigned int week,
+                                                    unsigned int year); 
+
+std::string createOptimizationFilename(const std::string& title,
+                                       std::shared_ptr<optPeriodAsString> opt_period_as_string,
+                                       const unsigned int optNumber,
+                                       const std::string& extension);
```
This function is only used in the implementation. IMO, it shouldn't be declared here, only in the cpp file.

guilpier-code replied:
I totally agree ! I will remove **createOptimizationFilename** from **filename.h** and keep it defined it in **filename.cpp**
==> done

guilpier-code replied:
By the way, in **filename.h**, there is an pure interface : **optPeriodAsString**.
This  leads to include **filename.h** where this abstraction is used, that is as argument of : 
- **OPT_AppelDuSimplexe(...)**
- **OPT_EcrireResultatFonctionObjectiveAuFormatTXT(...)**

This 2 functions are declared in **opt_fonctions.h**.
The problem is that today, changing **filename.h** implies changing **opt_fonctions.h**, that is recompiling a lot of code (as **opt_fonctions.h** is included in a lot of files).
I recommend isolating **optPeriodAsString** in a **opt-period-as-string.h** to be included in **opt_fonctions.h** and **filename.h**.
The abstraction **optPeriodAsString** is not supposed to change, but its implementations are supposed to change.
So, if we do that, changing  **filename.h** won't affect **opt_fonctions.h** anymore, and recompilation time will be much shorter.

guilpier-code replied:
Done

flomnes commented on src/solver/utils/filename.cpp:
```diff
@@ -1,23 +1,77 @@
 #include <sstream>
-#include <antares/study.h>
+#include <antares/logs/logs.h>
```
```suggestion
#include <antares/logs.h>
```

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/opt_optimisation_lineaire.cpp:
```diff
@@ -89,13 +100,23 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
-                                                           NumeroDeLIntervalle, 
-                                                           optimizationNumber);
+        {
+            double CoutOptimalSolution = OPT_ObjectiveFunctionResult(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber);
+            OPT_EcrireResultatFonctionObjectiveAuFormatTXT(CoutOptimalSolution, opt_period_as_string, optimizationNumber);
```
`CoutOptimalSolution` should be named `OptimalSolutionCost`

flomnes commented on src/solver/optimisation/opt_optimisation_lineaire.cpp:
```diff
@@ -35,6 +35,17 @@
 
+double OPT_ObjectiveFunctionResult(const PROBLEME_HEBDO* Probleme,
+                                   const int NumeroDeLIntervalle,
+                                   const int numeroOptimisation)
+{
+    if (numeroOptimisation == PREMIERE_OPTIMISATION)
```
```suggestion
                                   const int optimizationNumber)
{
    if (optimizationNumber == PREMIERE_OPTIMISATION)
```

guilpier-code replied:
Done

flomnes commented on src/solver/optimisation/opt_optimisation_lineaire.cpp:
```diff
@@ -89,13 +100,23 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
+        // An optimization period represents a sequence as <year>-<week> or <year>-<week>-<day>,
+        // depending whether the optimization is daily or weekly.
+        // These sequences are used when building the names of MPS or criterion files. 
+        std::shared_ptr<optPeriodAsString> opt_period_as_string = 
+            createOptimizationPeriod(ProblemeHebdo->OptimisationAuPasHebdomadaire,
```
```suggestion
        auto periodGenerator = 
            createOptimizationPeriod(ProblemeHebdo->OptimisationAuPasHebdomadaire,
```

guilpier-code replied:
Done

flomnes committed:
Fix MPS tests on Windows (#1045)

* Upload directory src/tests/mps in case of error

* Ignore CR;LF in Windows

* Revert spurious log

flomnes committed:
Merge remote-tracking branch 'github/develop' into fix/mps-files-when-daily-opt

flomnes committed:
Pass arg by const reference

guilpier-code committed:
[skip ci] MPS when daily optimization : corrections after review

guilpier-code committed:
[skip ci] MPS when daily optimization : corrections after review

flomnes commented on src/solver/optimisation/opt_optimisation_lineaire.cpp:
```diff
@@ -89,13 +100,22 @@ bool OPT_OptimisationLineaire(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace)
+                                                             ProblemeHebdo->weekInTheYear,
+                                                             ProblemeHebdo->year);
+
+        if (!OPT_AppelDuSimplexe(ProblemeHebdo, NumeroDeLIntervalle, optimizationNumber, opt_period_as_string))
```
optimizationRangeStringGenerator

guilpier-code replied:
Not sure **range** it is better than **period**

guilpier-code replied:
Done : **opt_period_as_string** ==> **optPeriodStringGenerator**

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,68 @@
+// ------------------------------------
+// Optimization period : base class
+// ------------------------------------
+class optPeriodAsString
+{
+public:
+    virtual std::string to_string() const = 0;
+    virtual ~optPeriodAsString() = default;
+};
```
Move this class declaration to another header file.

guilpier-code replied:
Done

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,68 @@
+// ------------------------------------
+// Daily optimization
+// ------------------------------------
+class optDailyAsString final : public optPeriodAsString
+{
+public:
+    optDailyAsString(unsigned int day, unsigned int week, unsigned int year);
+    std::string to_string() const override;
+
+private:
+    unsigned int day_ = 0;
+    unsigned int week_ = 0;
+    unsigned int year_ = 0;
+};
+
+
+// ------------------------------------
+// Weekly optimization
+// ------------------------------------
+class optWeeklyAsString final : public optPeriodAsString
+{
+public:
+    optWeeklyAsString(unsigned int week, unsigned int year);
+    std::string to_string() const override;
+
+private:
+    unsigned int week_ = 0;
+    unsigned int year_ = 0;
+};
```
Private

guilpier-code replied:
you mean private inheritance ?

flomnes replied:
These classes shouldn't be exposed in the public API. They should only be exposed inside a private header.

guilpier-code replied:
Done

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,68 @@
+    unsigned int year_ = 0;
+};
+
+std::shared_ptr<optPeriodAsString> createOptimizationPeriod(bool isOptimizationWeekly,
+                                                    unsigned int day,
+                                                    unsigned int week,
+                                                    unsigned int year);
```
Private

guilpier-code replied:
Done

flomnes commented on src/solver/utils/filename.h:
```diff
@@ -1,10 +1,68 @@
+                                                    unsigned int week,
+                                                    unsigned int year);
+
+std::string createCriterionFilename(std::shared_ptr<optPeriodAsString> opt_period_as_string, const unsigned int optNumber);
+std::string createMPSfilename(std::shared_ptr<optPeriodAsString> opt_period_as_string, const unsigned int optNumber);
```
Public

guilpier-code replied:
Done : nothing changed for these 2 functions

guilpier-code committed:
MPS when daily optimization : moving abstraction "optPeriodAsString" into a separate header

This to avoid a dependency to the filename.h's implementation code :
when changing filename.h's, implementation code, clients of "optPeriodAsString" do not recompile

flomnes commented:
TODO
- [ ] Update file **docs/reference-guide/13-file-format.md** to indicate this change in the output format in upcoming version 8.4.2
- [x] Resolve merge conflicts with the base branch

guilpier-code committed:
MPS when daily optimization : extracting optimization period as a string in .h/.cpp files

flomnes committed:
Merge branch 'fix/refactor-export-structure' into fix/mps-files-when-daily-opt

flomnes committed:
Restore function getFilenameWithExtension

guilpier-code committed:
MPS when daily optimization : renaming "opt period as string" into "opt period string generator"

guilpier-code committed:
[skip ci] MPS when daily optimization : renaming a local variable

guilpier-code committed:
[skip ci] MPS when daily optimization : some more renaming (classes and a local variable)

flomnes committed:
[skip ci] Rename opt_period_as_string -> optPeriodStringGenerator
